packages/cardano-services/src/util/polling/EpochPollService.ts

```diff
@@ -0,0 +1,64 @@
+import { AsyncAction, InMemoryCache, UNLIMITED_CACHE_TTL } from '../../InMemoryCache';
+
+export const EPOCH_POLL_INTERVAL_DEFAULT = 10_000;
+export const CURRENT_EPOCH_CACHE_KEY = 'current_epoch';
+
+/**
+ * Class to handle epoch rollover through db polling
+ */
+export class EpochPollService {
+  static #instance: EpochPollService;
+  #cache: InMemoryCache;
+  #timeoutId: number;
+  #interval: number;
+  #asyncAction: AsyncAction<number>;
+
+  private constructor(cache: InMemoryCache, asyncAction: AsyncAction<number>, interval: number) {
+    this.#cache = cache;
+    this.#interval = interval;
+    this.#asyncAction = asyncAction;
+    this.executePoll = this.executePoll.bind(this);
+  }
+
+  /**
+   * Poll execution to determine new epoch rollover
+   */
+  private async executePoll() {
+    const lastEpoch = await this.#asyncAction();
+    const currentEpoch = this.#cache.getVal<number>(CURRENT_EPOCH_CACHE_KEY);
+    const shouldClearCache = !!(currentEpoch && lastEpoch > currentEpoch);
+
+    if (!currentEpoch || shouldClearCache) {
+      shouldClearCache ? this.#cache.clear() : void 0;
+      this.#cache.set<number>(CURRENT_EPOCH_CACHE_KEY, lastEpoch, UNLIMITED_CACHE_TTL);
+    }
+  }
+
+  /**
+   * Creates a single instance of EpochPollService
+   *
+   * @returns {EpochPollService} EpochPollService instance
+   */
+  static create(cache: InMemoryCache, asyncAction: AsyncAction<number>, interval: number): EpochPollService {
```

Comment:
Since it takes in an asyncAction and interval, singleton doesn't make much sense here. It should at least group instances by `asyncAction`. I suggest to not do singleton and just inject the same instance as dependency to every dependent.

Reply:
The `asyncAction` was one of the providers' builder queries, which at the point of poll service `create` the builder was not yet created. The idea was each of the providers to know about its asyncAction only, but since it's a common query then could be moved into epoch poll service scope/directory and passed during poll service init/start in `loadHttpServer.ts`

Reply:
Another thing is that `db` could be `undefined` at point of `epochPollService` initialization if we create one instance in [loadHttpServer](https://github.com/input-output-hk/cardano-js-sdk/blob/419252a9082b710ad42012dd4c01cc397de8e658/packages/cardano-services/src/Program/loadHttpServer.ts#L134) and pass it down to every dependant. So we might want to benefit from `withDb()` decorator and instantiate epoch service within each dependant service map [StakePool](https://github.com/input-output-hk/cardano-js-sdk/blob/419252a9082b710ad42012dd4c01cc397de8e658/packages/cardano-services/src/Program/loadHttpServer.ts#L87-L93) and [NetworkInfo](https://github.com/input-output-hk/cardano-js-sdk/blob/419252a9082b710ad42012dd4c01cc397de8e658/packages/cardano-services/src/Program/loadHttpServer.ts#L107-L122) but then the singleton helps in that case? Any thoughts @mkazlauskas

Reply:
The changes are applied under https://github.com/input-output-hk/cardano-js-sdk/pull/370/commits/7c8cd1f53bc0bfecbb7fe3af6c33fa0b409288f6 , as we agreed - use `memoize` for single instance creation of `DbSyncEpochPollService` on each dependant service load method.